src/ray/pubsub/publisher.cc

```diff
@@ -278,28 +286,33 @@ bool SubscriberState::PublishIfPossible(bool force_noop) {
 
   // No message should have been added to the reply.
   RAY_CHECK(long_polling_connection_->reply->pub_messages().empty());
+
   if (!force_noop) {
-    for (int i = 0; i < publish_batch_size_ && !mailbox_.empty(); ++i) {
-      const rpc::PubMessage &msg = *mailbox_.front();
+    for (auto it = mailbox_.begin(); it != mailbox_.end(); it++) {
+      if (long_polling_connection_->reply->pub_messages().size() >= publish_batch_size_) {
+        break;
+      }
+      const rpc::PubMessage &msg = **it;
       // Avoid sending empty message to the subscriber. The message might have been
       // cleared because the subscribed entity's buffer was full.
       if (msg.inner_message_case() != rpc::PubMessage::INNER_MESSAGE_NOT_SET) {
         *long_polling_connection_->reply->add_pub_messages() = msg;
       }
-      mailbox_.pop();
     }
   }
+
   long_polling_connection_->send_reply_callback(Status::OK(), nullptr, nullptr);
 
   // Clean up & update metadata.
   long_polling_connection_.reset();
+  // Clean up & update metadata.
   last_connection_update_time_ms_ = get_time_ms_();
   return true;
 }
 
 bool SubscriberState::CheckNoLeaks() const {
   // If all message in the mailbox has been replied, consider there is no leak.
-  return !long_polling_connection_ && mailbox_.empty();
```

Comment:
Why do we remove this?

Reply:
- this is only used for testing, and there is already another helper function `ConnectionExists` does't connection check.
- we also changed test code to call `ConnectionExists` accordingly.